lib/omnibus/s3_license_cache.rb

```diff
@@ -162,16 +162,18 @@ def key_for(software, license_file)
           raise InsufficientSpecification.new(:version, software)
         end
 
-        unless software.hash
-          raise InsufficientSpecification.new(:hash, software)
-        end
-
-        # We add Software#shasum in the cache key. It's an accurate way
+        # We add a custom hash of the software recipe in the cache key. It's an accurate way
         # to know if a software definition changed, as it takes into account the
         # resolved version (ie. the git commit hash if the source is a git repository,
         # the hashsum of the downloaded file if the source is a remote file), the project,
         # and all build commands run in the software definition.
-        "licenses/#{software.name}-#{software.version}-#{software.shasum}/#{File.basename(license_file)}"
+        # We can't rely on the software.shasum directly as it includes the hash of all the other
+        # softwares it depends on, which causes the checksum to change way too often.
```

Comment:
Let's update this now that we have found the real culprit?

`software.shasum` includes `project.shasum`, which includes `install_dir` ; and since the OCI projects have a different `install_dir` every run by design, this causes them to always upload licenses, filling our cache with garbage, ultimately making all cache lookups significantly slower.

Reply:
Why does the size of the cache affect lookup performance? I'd expect each file to be addressed directly on retrieval. I can see how a problem could exist in terms of not being able to benefit from the cache on OCI builds (the same way it currently happens with the git cache) but that's not being mentioned directly anywhere here.

Reply:
Because of this part: https://github.com/DataDog/omnibus-ruby/blob/1a5836bb3dc3254f47914712758541d74aa4b60f/lib/omnibus/s3_cache.rb#L45-L47.

For cache operations, omnibus works in a weird way. Instead of computing each software's cache key and querying S3 once for the presence of this cache key, it prefers first computing the whole list of elements that are in the S3 bucket (to only do one network operation, I guess), and then does cache presence lookups against this local list.

The cost of this operation is linear in the number of elements in the cache. At 600k elements (what we are at right now, with an average of +5k/day), this takes almost 4 minutes.

Reply:
Comment updated

Reply:
That makes a lot of sense, thanks for the clarification. Yeah, I can see that implementation decision make sense for reasonable numbers of elements (within the range of what a list objects request can give you); and it now makes total sense that it's causing a performance issue. Plus it explains the ever increasing count of objects in the bucket (that stopped correlating with the use of bucket for git caching).